stats.lua: add keybindings info page

avih opened the pull request:
- Move the internal performance info from page 4 to page 0.
- Add key-bindings list at page 4.
- Support help-like printout of the whole list to the terminal.

The alignment/grouping style is heavily inspired by @medhefgo's #8924 but otherwise it's an independent implementation.

<details>
  <summary>Inside: terminal output listing for --no-config</summary>

By default the key names and section titles are in bold, but using `-yes` instead of `yes` disables the terminal escape sequences. The interactive video-window (libass) output is similar in style.

```
$ mpv --idle --no-config --script-opts=stats-help-keys=-yes
Active key bindings:
  
                    audio 
             SHARP  cycle audio
            Ctrl++  add audio-delay 0.100
            Ctrl+-  add audio-delay -0.100
  
                    brightness 
                 3  add brightness -1
                 4  add brightness 1
  
                    chapter 
                 !  add chapter -1
                 @  add chapter 1
              PGUP  add chapter 1
             PGDWN  add chapter -1
  
                    contrast 
                 1  add contrast -1
                 2  add contrast 1
  
                    deinterlace 
                 d  cycle deinterlace
  
                    edition 
                 E  cycle edition
  
                    frame 
                 ,  frame-back-step
                 .  frame-step
  
                    fullscreen 
                 f  cycle fullscreen
               ESC  set fullscreen no
     MBTN_LEFT_DBL  cycle fullscreen
  
                    gamma 
                 5  add gamma -1
                 6  add gamma 1
  
                    hwdec 
            Ctrl+h  cycle-values hwdec "auto" "no"
  
                    loop 
                 l  ab-loop
                 L  cycle-values loop-file "inf" "no"
  
                    mute 
                 m  cycle mute
              MUTE  cycle mute
  
                    ontop 
                 T  cycle ontop
  
                    osd 
                 O  no-osd cycle-values osd-level 3 1
  
                    panscan 
                 e  add panscan +0.1
                 w  add panscan -0.1
                 W  add panscan +0.1
  
                    pause 
                 p  cycle pause
              PLAY  cycle pause
             PAUSE  cycle pause
             SPACE  cycle pause
          PLAYONLY  set pause no
         PAUSEONLY  set pause yes
         PLAYPAUSE  cycle pause
        MBTN_RIGHT  cycle pause
  
                    playlist 
                 <  playlist-prev
                 >  playlist-next
              NEXT  playlist-next
              PREV  playlist-prev
             ENTER  playlist-next
         MBTN_BACK  playlist-prev
      MBTN_FORWARD  playlist-next
  
                    quit 
                 q  quit
                 Q  quit-watch-later
              STOP  quit
             POWER  quit
         CLOSE_WIN  quit
            Ctrl+c  quit 4
            Ctrl+w  quit
  
                    saturation 
                 7  add saturation -1
                 8  add saturation 1
  
                    screenshot 
                 s  screenshot
                 S  screenshot video
             Alt+s  screenshot each-frame
            Ctrl+s  screenshot window
  
                    script: console 
                 `  script-binding console/enable
  
                    script: osc 
               DEL  script-binding osc/visibility
  
                    seek 
                UP  seek  60
              DOWN  seek -60
              LEFT  seek -5
             RIGHT  seek  5
            REWIND  seek -60
           FORWARD  seek 60
          WHEEL_UP  seek 10
        WHEEL_DOWN  seek -10
          Shift+BS  revert-seek
          Shift+UP  no-osd seek  5 exact
        Shift+DOWN  no-osd seek -5 exact
        Shift+LEFT  no-osd seek -1 exact
        Shift+PGUP  seek 600
       Shift+PGDWN  seek -600
       Shift+RIGHT  no-osd seek  1 exact
     Shift+Ctrl+BS  revert-seek mark
  
                    show 
                 o  show-progress
                 P  show-progress
                F8  show_text ${playlist}
                F9  show_text ${track-list}
  
                    speed 
                 [  multiply speed 1/1.1
                 ]  multiply speed 1.1
                 {  multiply speed 0.5
                 }  multiply speed 2.0
                BS  set speed 1.0
  
                    sub 
                 F  add sub-scale -0.1
                 G  add sub-scale +0.1
                 j  cycle sub
                 J  cycle sub down
                 r  add sub-pos -1
                 R  add sub-pos +1
                 t  add sub-pos +1
                 u  cycle-values sub-ass-override "force" "no"
                 v  cycle sub-visibility
                 V  cycle sub-ass-vsfilter-aspect-compat
                 x  add sub-delay +0.1
                 z  add sub-delay -0.1
                 Z  add sub-delay +0.1
             Alt+v  cycle secondary-sub-visibility
         Ctrl+LEFT  no-osd sub-seek -1
        Ctrl+RIGHT  no-osd sub-seek  1
   Shift+Ctrl+LEFT  sub-step -1
  Shift+Ctrl+RIGHT  sub-step 1
  
                    video 
                 _  cycle video
                 A  cycle-values video-aspect-override "16:9" "4:3" "2.35:1"...
             Alt++  add video-zoom   0.1
             Alt+-  add video-zoom  -0.1
            Alt+BS  set video-zoom 0 ; set video-pan-x 0 ; set video-pan-y 0
            Alt+UP  add video-pan-y  0.1
          Alt+DOWN  add video-pan-y -0.1
          Alt+LEFT  add video-pan-x  0.1
         Alt+RIGHT  add video-pan-x -0.1
  
                    volume 
                 *  add volume 2
                 /  add volume -2
                 0  add volume 2
                 9  add volume -2
         VOLUME_UP  add volume 2
        WHEEL_LEFT  add volume -2
       VOLUME_DOWN  add volume -2
       WHEEL_RIGHT  add volume 2
  
                    window 
             Alt+0  set window-scale 0.5
             Alt+1  set window-scale 1.0
             Alt+2  set window-scale 2.0
```
</details>

avih commented:
Force pushed the following changes to the relevant commits:
- Fixed script error at `keybinding_info()` if `get_kbinfo_lines()` is called before `eval_ass_formatting()` was called at least once (swap order) - issue found by @guidoc.
- Changed the `while` loop at `cmd_subject()` to a nicer `repeat`-`until` loop - suggested by @guidoc.
- DOCS: mention that `--idle` is needed for `stats-help-keys`.
- Comments: changed "regex" to "pattern" (@TheAMM), fix typo "soace" -> "space".

guidocella commented on player/lua/stats.lua:
```diff
@@ -995,3 +1137,22 @@ mp.register_event("video-reconfig",
+
+--  --script-opts-add=stats-help-keys=[-]{yes|<TERM-WIDTH>}
+if o["help-keys"] ~= "no" then
+    mp.command("no-osd set really-quiet yes")
```
This can be shortened to

```suggestion
    mp.set_property("really-quiet", "yes")
```

avih commented:
Additional commits of one nice improvement of the string processing for the grouping titles (can now parse quotes, and also more efficient), and few minor improvements - display binding comment too if available, include non-interactive stats.lua keys at the list, minor docs enhancement.

These will be squashed later to the "add stats page 4" commit.

Argon- commented on player/lua/stats.lua:
```diff
@@ -77,6 +78,8 @@ local o = {
     no_ass_it1 = "\027[3m",
     no_ass_it0 = "\027[0m",
+
+    ["help-keys"] = "no",  -- print the list on startup and quit mpv
```
Is there a reason why this uses `-` instead of `_` as the other options do? There is something about this at line 1144 but I don't get it.
`print the list` -> what list? Is it the same output otherwise printed on screen?

avih replied:
> Is there a reason why this uses - instead of _ as the other options do?

Other options use "-", and for users it's way more convenient to use "-" instead of "_" (it annoys me to no end when I need to use "_" for words separation in script opts).

Only script options usually use "_", and only because it's more convenient to use at the code itself (can be used as `opts.foo_bar` instead of the more awkward `opts["foo-bar"]`.

> There is something about this at line 1144 but I don't get it.

Not sure I understand. Please be more specific.

> `print the list` -> what list? Is it the same output otherwise printed on screen?

Yes. Same output which can be browsed interactively at the video window or at the terminal OSD. It can be printed with the same form to the terminal like a help page.

avih replied:
> `print the list` -> what list?

The first comment has some collapsed content. Expand it to see how it looks at the terminal.

Argon- replied:
While I understand your sentiment regarding dash/underscore options, it's a personal preference and we don't even know if your subjective one is representative for other users. However, this makes the stats script options inconsistent with a mix of dash and underscore options. I would really prefer to avoid that.
 
(Note: personally I have no preference other than consistency)

avih replied:
> However, this makes the stats script options inconsistent with a mix of dash and underscore options. I would really prefer to avoid that.

To me using mixed dash/underscore is inconsistent on its own, and using dashes-only is consistent on its own and also consistent with general mpv options.

I definitely understand where you're comming from, but I still strongly prefer dashes-only.

Maybe others could share their preference?

guidocella replied:
I think `keys_help` is more consistent because all `script-opts` use _. If i were to type this on the terminal from memory months from now, I would use _ and expect it to work.
I also find that using different symbols for separating the script and its options and for the options themselves helps to differentiate them visually.

avih replied:
> I think `keys_help` is more consistent because all `script-opts` use _

No, they use both. For instance `--script-opts=ytdl_hook-ytdl_path=whatever`, and It's a damn pain to remember which are `-` and which are `_`. Personally I almost never get the dashes/underscores correct on first attempt.

Some of the underscores are sort of forced, because mpv convers script-names to use underscores (that's the `ytdl_hook` part). Then there's always a dash, and then the actual options are usually with underscore - for no good reason other than it's more convinient to write at the source code `o.foo_bar` instead of `o["foo-bar"]`.

So it ends up that if the script name is one word and the option name is one word, then it always uses only dashes. This is very common, for instance `--script-opts=osc-visibility=always` or `--script-opts=stats-font=whatever`.

However, script name and/or option names which are more than one word are separated with underscore.

Of the builtin scripts, only `ytdl_hook` is two words, and many of the option-names are one word, so in practice many of script opts of most builtin scripts are dashes-only.

Personally, I really dislike the forced mixed dash/underscores. However:
- If my opinion is a minority then I'll cave.
- On such case, I'd prefer to find one word so that you guys feel it's consistent with other script-opts. Maybe `--script-opts=stats-keys=help` or some such?

Argon- replied:
> If my opinion is a minority then I'll cave.

I don't think if this should be about opinions, for me this is about consistency. You demonstrated how the current system as a whole is inconsistent and I give you that but at least the current stats script-options are consistent and use only underscores. I would really prefer not to include an unnecessary and additional inconsistency here now.

Now, changing _all_ of stats' script-options from underscores to dashes is a possibility. As said before, I have no personal preference here, so I'd be fine to change all options to dashes if that helps with consistency. However, this would render all existing `script-opts/stats.conf` invalid and, therefore, should imo only be done when there's a strong majority among users in favor of this change.

avih replied:
> I don't think if this should be about opinions, for me this is about consistency.

Yeah, I see your point, but I think differently _on this specific case_. I'm fine with doing what most think is best. No hard feeelings here.

> Now, changing _all_ of stats' script-options from underscores to dashes is a possibility.

Nah, too painful for users and for anyone who might try that. IMHO the only way to support consistent dashes is to automate it so that both would work for script-opts - like they are for mpv command-names (dashes and underscores are interchangable). This way no script needs to be modified, and nothing breaks (or very little breaks where they conclist), and everyone is happy.

However, that's not for this PR.

So, you don't want dash-separated option names, and I don't want underscores.

How about you help me find one word so that it doesn't need to use underscores, and we're both happy?

What about `--script-opts=stats-helpkeys=yes` ? or `stats-printkeys=yes`? (combined words are very common at the `osc.lua` for instance) or `--script-opts=stats-keys=help`? Or something else which you have in mind?

avih replied:
Also, if you look at the commit message which added this option, you'll see that I fully acknowledged the awkwardness of the invocation, though not due to the dashes (this actually makes it better than underscores), but rather because it's basically a help page which is triggered via script-opts because that's what scripts can use.

I mentioned that in the future we can try to streamline the invocation, and @guidocella suggested something like `--input-bindings` or `input-keylist` (similar to the existing `--list-protocols` and `--input-cmdlist`). That would definitely be better than the approach of this PR.

However, it also doesn't contradict. We can have something now via script-opts to make it useful, and when we find the time to add more streamlined invocation, we'll do that too.

avih replied:
> What about ...

How about `--script-opts=stats-bindlist=yes` in similar naming as the existing `--input-cmdlist` and `--input-keylist` ?

I'm going to assume this is acceptable unless there's an explicit objection (at which case do suggest an altrenative using one word).

avih commented:
> I would really prefer to avoid .. (using dash as name-separator for two-words stats options).

Changed the option `keys-help` to `bindlist` (`--script-opts=stats-bindlist=yes`), and squashed into the relevant commit.

That's only one word now, and in similar naming scheme as the existing `--input-keylist` and `--input-cmdlist`.

> Additional commits of one nice improvement of the string processing for the grouping titles (can now parse quotes, and also more efficient), and few minor improvements - display binding comment too if available, include non-interactive stats.lua keys at the list, minor docs enhancement.

Squashed into the "add page 4" commit, together with an additional streamline of the grouping function (handle the name prefixes more efficiently - similar to the token prefixes).

I'll push this to master soon if there are no further comments.

Argon- commented:
I'm fine with `bindlist` :)

avih commented:
Thanks for the feedback.

avih commented:
> include non-interactive stats.lua keys at the list

Oops, this change accidentally added the interactive stats keys too, e.g. for 4 it ended up listed as `script-binding stats/4` rather than `add brightness 1` or some such.

Fixed in master - fab25ac004be0de84d5f70f16499a73ae14b0f6d
